### PR TITLE
Consolidate contribution documentation

### PR DESCRIPTION
**Pull request by tristan957**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -251,3 +413,14 @@ meson compile -C build
...
+### Code Formatting
+
+Proper formatting is required for passing continuous integration checks.
+
+#### C
+
+HSE uses `clang-format` for formatting C code. We include a script at
+[`scripts/dev/clang-format.sh`](./scripts/dev/clang-format.sh) that you can
+use to format either the whole tree or certain files. It is recommended that you
+just hook up your editor to format files for you however.
```
Comment (alexttx) [inline CONTRIBUTING.md:426]: Maybe mention that the script can be used to check formatting.

**Comment by alexttx** [comment]
Remove this line about the  PR template:

* Use the predefined PR template and specify which issue the commit addresses, what the commit does, and provide a concise description of the change.

**Comment by alexttx** [comment]
Seems like we can remove the section on how to clone.  Anybody that's gonna want source should know how to clone the repo.

**Comment by alexttx** [comment]
Other suggestions:
- Consolidate `Coding Style` and `Code Formatting` sections.
- Split `Testing` into two sections: `Running Tests`, and `Developing Tests`.  Put the latter at the end of the document to hide the noise of that section.  Move the `Coverage` section into the new `Developing Tests`.
- The info for distribution maintainers is spread among three sections (Building For Dist Maintainers, Distributing the Source, and Distributing Binaries).  Move all that into one section name "Distribution Maintainers"  or something similar.

**Review by alexttx — APPROVED** [review]

**Comment by tristan957** [comment]
Perhaps the section for distribution maintainers should go elsewhere. It isn't like they are going to be reading the CONTRIBUTING.md document anyways. README.md seems like the next logical place, but I don't know.

I don't think Distributing either the source or the binaries does much good for distribution maintainers. I've done packaging for distros before, and that section isn't valuable to me. I am not sure who that section would apply to other than us. Maybe a MAINTAINERS.md? Not sure.

**Comment by alexttx** [comment]
You know more about dist packaging that anybody else our our team, so I'll let you make the call.  If it has not value, then lose  the text.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -150,6 +152,185 @@ meson test -C build [tests...]
...
+###### `multiple definition of mtfm_foo_foo_create_getreal`
+
+More than on source file has `#include "cn_foo_ut_impl.i"`.
+
+###### `undefined reference to mtfm_foo_foo_create_get`
+
+Could indicate one of two things:
+
+* The source file with `foo_create()` is not compiled
+* The source file with `foo_create()` is missing `#include "cn_foo_ut_impl.i"`
```
Comment (beckerg) [inline CONTRIBUTING.md:327]: I think there are some other ways to manifest this that require rebuilding from a clean slate to fix it (I suspect due to something broken regarding dependencies).

**Review by beckerg — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (tristan957) [reply CONTRIBUTING.md:327]: I remember seeing this in the past, but not anymore. If you do run into it, let me know since it might be a Meson issue :).

**Review by smoyergh — APPROVED** [review]
Looks good, though I cannot review or comment on the mocking docs.

**Comment by beckerg** [comment]
You're not alone Steve!

**Pull request merged**